projects/components/src/filtering/filter/filter.service.ts

```diff
@@ -0,0 +1,205 @@
+import { Injectable } from '@angular/core';
+import { NavigationService } from '@hypertrace/common';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { FilterBuilderLookupService } from './builder/filter-builder-lookup.service';
+import { AbstractFilterBuilder } from './builder/types/abstract-filter-builder';
+import { Filter, IncompleteFilter } from './filter';
+import { FilterAttribute } from './filter-attribute';
+import { FilterOperator, fromUrlFilterOperator, toUrlFilterOperator } from './filter-operators';
+import { FilterParserLookupService } from './parser/filter-parser-lookup.service';
+import { SplitFilter } from './parser/parsed-filter';
+import { AbstractFilterParser } from './parser/types/abstract-filter-parser';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FilterService {
+  private static readonly FILTER_QUERY_PARAM: string = 'filter';
+
+  public constructor(
+    private readonly navigationService: NavigationService,
+    private readonly filterBuilderLookupService: FilterBuilderLookupService,
+    private readonly filterParserLookupService: FilterParserLookupService
+  ) {}
+
+  /*
+   * URL Filtering
+   */
+
+  public getUrlFiltersChanges$(attributes: FilterAttribute[]): Observable<Filter[]> {
+    return this.navigationService.navigation$.pipe(map(() => this.getUrlFilters(attributes)));
+  }
+
+  public getUrlFilters(attributes: FilterAttribute[]): Filter[] {
+    return this.navigationService
+      .getAllValuesForQueryParameter(FilterService.FILTER_QUERY_PARAM)
+      .map(filterString => this.parseUrlFilterString(attributes, filterString))
+      .filter((filter): filter is Filter => filter !== undefined);
+  }
+
+  public setUrlFilters(filters: Filter[]): void {
+    this.navigationService.addQueryParametersToUrl({
+      [FilterService.FILTER_QUERY_PARAM]: filters.length === 0 ? undefined : filters.map(f => f.urlString)
+    });
+  }
+
+  public applyUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
+    const filters = this.getUrlFilters(attributes);
+
+    const foundIndex = filters.findIndex(f => filter.field === f.field);
+
+    if (foundIndex !== -1) {
```

Comment:
Isn't this an explicit implementation of the behavior we currently have an outstanding bug for? we're replacing rather than adding if it's an existing field.

Reply:
Will follow up with those changes separately. Trying to keep this PR focused. Its big enough as is.